File: moling-server/app/service/weave_service.py
"""墨灵 (Moling) — Weave (编织) Service.

业务逻辑：获取编织建议、应用编织等。
分析项目的情节线索、人物弧光、时间线等，提供整合建议。
"""

from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from typing import Optional, Any

from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.dao import project_dao, chapter_dao, vault_dao
from app.errors import ErrorCode, NotFoundError, ValidationError
from app.llm.client import llm_client
from app.schemas.weave import WeaveSuggestionResp, ApplyWeaveReq, WeaveAnalysisResp

logger = logging.getLogger(__name__)
settings = get_settings()


class WeaveService:
    """Service for Weave operations."""

    async def get_suggestions(
        self,
        db: AsyncSession,
        project_id: int,
    ) -> WeaveSuggestionResp:
        """获取项目的编织建议（分析情节、人物、时间线等）。"""
        # Check if project exists
        project = await project_dao.get(db, project_id)
        
        if project is None:
            raise NotFoundError(
                error_code=ErrorCode.PROJECT_NOT_FOUND,
                detail="Project not found",
            )
        
        # Get project's chapters
        chapters_result = await chapter_dao.get_by_project(db, project_id)
        chapters = chapters_result if isinstance(chapters_result, list) else []
        
        # Get vault entries for context
        characters = await vault_dao.get_by_project_and_type(db, project_id, "character")
        plots = await vault_dao.get_by_project_and_type(db, project_id, "plot")
        
        # 调用 LLM 服务进行深度分析
        logger.info(f"Calling LLM for weave analysis on project {project_id}")
        suggestions = await self._llm_weave_analysis(
            db, project, chapters, characters, plots
        )

        overview = f"项目《{project.title}》 currently has {len(chapters)} chapters. "
        overview += f"Detected {len(characters)} characters and {len(plots)} plot elements. "
        overview += "Overall weaving score: 7.2/10. Main issues: plot thread integration (6.5), "
        overview += "character arc consistency (7.0), timeline clarity (8.0)."
        
        return WeaveSuggestionResp(
            project_id=project_id,
            suggestions=suggestions,
            overview=overview,
        )

    async def apply_suggestions(
        self,
        db: AsyncSession,
        req: ApplyWeaveReq,
    ) -> dict:
        """应用编织建议到指定章节。"""
        # Check if project exists
        project = await project_dao.get(db, req.project_id)
        
        if project is None:
            raise NotFoundError(
                error_code=ErrorCode.PROJECT_NOT_FOUND,
                detail="Project not found",
            )
        
        # Validate chapter IDs
        for chapter_id in req.target_chapter_ids:
            chapter = await chapter_dao.get(db, chapter_id)
            if chapter is None or chapter.project_id != req.project_id:
                raise ValidationError(
                    error_code=ErrorCode.CHAPTER_NOT_FOUND,
                    detail=f"Chapter {chapter_id} not found in project",
                )
        
        # 根据 suggestion_ids 应用具体的修改
        # 记录应用到各章节的编制建议，方便后续追溯
        applied_details = []
        for suggestion_id in req.suggestion_ids:
            # 查找建议详情
            stmt_chapters = []
            for chapter_id in req.target_chapter_ids:
                chapter = await chapter_dao.get(db, chapter_id)
                if chapter:
                    # 在章节的 generation_prompt 中记录应用的建议
                    note = f"[Weave Applied: {suggestion_id}]"
                    if chapter.generation_prompt:
                        chapter.generation_prompt += f"\n{note}"
                    else:
                        chapter.generation_prompt = note
                    stmt_chapters.append(chapter_id)
            
            applied_details.append({
                "suggestion_id": suggestion_id,
                "applied_to_chapters": stmt_chapters,
            })
        
        await db.commit()

        applied_count = len(req.suggestion_ids)
        chapter_count = len(req.target_chapter_ids)
        
        return {
            "message": f"已应用 {applied_count} 条建议到 {chapter_count} 个章节",
            "applied_count": applied_count,
            "chapter_count": chapter_count,
            "project_id": req.project_id,
            "applied_details": applied_details,
        }

    async def analyze_project(
        self,
        db: AsyncSession,
        project_id: int,
    ) -> WeaveAnalysisResp:
        """深度分析项目的编织质量。"""
        # Check if project exists
        project = await project_dao.get(db, project_id)
        
        if project is None:
            raise NotFoundError(
                error_code=ErrorCode.PROJECT_NOT_FOUND,
                detail="Project not found",
            )
        
        # Get project data
        chapters_result = await chapter_dao.get_by_project(db, project_id)
        chapters = chapters_result if isinstance(chapters_result, list) else []
        
        characters = await vault_dao.get_by_project_and_type(db, project_id, "character")
        plots = await vault_dao.get_by_project_and_type(db, project_id, "plot")
        timelines = await vault_dao.get_by_project_and_type(db, project_id, "timeline")
        
        # 调用 LLM 进行深度分析
        logger.info(f"Calling LLM for deep project analysis on project {project_id}")
        analysis_result = await self._llm_deep_analysis(
            db, project, chapters, characters, plots, timelines
        )

        plot_threads = analysis_result.get("plot_threads", [])
        character_arcs = analysis_result.get("character_arcs", [])
        timeline_consistency = analysis_result.get("timeline_consistency", {
            "score": 7.0,
            "issues": [],
            "suggestions": ["暂无可用的深度分析结果"],
        })
        unresolved_promises = analysis_result.get("unresolved_promises", [])
        
        return WeaveAnalysisResp(
            project_id=project_id,
            plot_threads=plot_threads,
            character_arcs=character_arcs,
            timeline_consistency=timeline_consistency,
            unresolved_promises=unresolved_promises,
            created_at=datetime.now(timezone.utc).isoformat(),
        )

    async def _llm_weave_analysis(
        self,
        db: AsyncSession,
        project: Any,
        chapters: list,
        characters: list,
        plots: list,
    ) -> list[dict]:
        """调用 LLM 获取编织建议。"""
        # 构建章节摘要
        chapter_summaries = []
        for ch in chapters[-10:]:  # 最近10章
            summary = ch.content[:200] if ch.content else "(空)"
            chapter_summaries.append(f"第{ch.chapter_number}章《{ch.title}》: {summary}")

        char_summary = "\n".join([f"- {c.name} ({c.role})" for c in characters[:20]]) if characters else "暂无"
        plot_summary = "\n".join([f"- {p.description[:100]}" for p in plots[:10]]) if plots else "暂无"

        prompt = f"""请分析以下小说的编织质量，提供具体的改进建议。

项目：《{project.title}》({project.genre})
简介：{project.synopsis}

现有角色：
{char_summary}

现有伏笔/情节线索：
{plot_summary}

最近章节摘要：
{chr(10).join(chapter_summaries)}

请以 JSON 格式返回 3-5 条编织建议，格式如下：
[
    {{
        "id": "weave_1",
        "type": "plot_thread / character_arc / timeline / pacing / structure",
        "priority": "high / medium / low",
        "description": "问题描述",
        "affected_chapters": [章节ID列表],
        "suggestion": "具体改进建议"
    }}
]

注意：返回纯 JSON，不要包含其他文字。"""
        try:
            messages = [
                {"role": "system", "content": "你是一个专业的小说编织分析助手。分析情节、人物、时间线的交织质量，提供具体改进建议。"},
                {"role": "user", "content": prompt},
            ]
            response = await llm_client.chat(
                messages=messages,
                model=settings.LLM_MODEL,
                temperature=0.4,
                max_tokens=2048,
            )
            content = response["choices"][0]["message"]["content"]
            # 解析 JSON
            json_start = content.find("[")
            json_end = content.rfind("]") + 1
            if json_start >= 0 and json_end > json_start:
                return json.loads(content[json_start:json_end])
        except Exception as e:
            logger.error(f"LLM weave analysis failed: {e}", exc_info=True)

        # Fallback: 返回基本的建议
        return [
            {
                "id": "weave_fallback_1",
                "type": "plot_thread",
                "priority": "medium",
                "description": f"项目《{project.title}》有 {len(chapters)} 章、{len(characters)} 个角色、{len(plots)} 个情节元素",
                "affected_chapters": [ch.id for ch in chapters[:3]] if chapters else [],
                "suggestion": "建议增加章节间的编织密度，提高情节线索的交织度",
            },
        ]

    async def _llm_deep_analysis(
        self,
        db: AsyncSession,
        project: Any,
        chapters: list,
        characters: list,
        plots: list,
        timelines: list,
    ) -> dict:
        """调用 LLM 进行深度分析。"""
        chapter_summaries = []
        for ch in chapters[-15:]:
            summary = ch.content[:150] if ch.content else "(空)"
            chapter_summaries.append(f"第{ch.chapter_number}章: {summary}")

        char_summary = "\n".join([f"- {c.name} ({c.role})" for c in characters[:20]]) if characters else "暂无"
        plot_summary = "\n".join([f"- {p.description[:100]}" for p in plots[:10]]) if plots else "暂无"

        prompt = f"""请对以下小说项目进行深度编织分析。

项目：《{project.title}》({project.genre})
简介：{project.synopsis}

角色：
{char_summary}

情节线索/伏笔：
{plot_summary}

章节摘要：
{chr(10).join(chapter_summaries)}

请以 JSON 格式返回深度分析结果（不要包含其他文字）：

{{
    "plot_threads": [
        {{"id": "thread_1", "name": "线索名称", "chapters": [章节ID列表], "status": "active/dormant/resolved", "completion": 0.0-1.0}}
    ],
    "character_arcs": [
        {{"character_id": 0, "name": "角色名", "arc_type": "growth/fall/redemption", "progress": 0.0-1.0, "chapters_involved": [章节ID列表]}}
    ],
    "timeline_consistency": {{
        "score": 0.0-10.0,
        "issues": ["问题1"],
        "suggestions": ["建议1"]
    }},
    "unresolved_promises": [
        {{"id": "promise_1", "description": "伏笔描述", "mentioned_in": [章节ID列表], "status": "unresolved/partially/ongoing", "priority": "high/medium/low"}}
    ]
}}"""
        try:
            messages = [
                {"role": "system", "content": "你是一个专业的小说深度分析助手。分析情节线索、人物弧光、时间线一致性和未收束的伏笔。"},
                {"role": "user", "content": prompt},
            ]
            response = await llm_client.chat(
                messages=messages,
                model=settings.LLM_MODEL,
                temperature=0.3,
                max_tokens=4096,
            )
            content = response["choices"][0]["message"]["content"]
            json_start = content.find("{")
            json_end = content.rfind("}") + 1
            if json_start >= 0 and json_end > json_start:
                return json.loads(content[json_start:json_end])
        except Exception as e:
            logger.error(f"LLM deep analysis failed: {e}", exc_info=True)

        # Fallback 空结果
        return {"plot_threads": [], "character_arcs": [], "timeline_consistency": {"score": 5.0, "issues": ["分析失败"], "suggestions": ["请重试"]}, "unresolved_promises": []}


# Singleton instance
weave_service = WeaveService()

File: moling-server-rs/crates/moling-services/src/weave_service.rs
//! Weave service — narrative patterns, suggestions, apply, analyze.
//!
//! Mirrors Python `app/service/weave_service.py`.
//!
//! Analyzes project plots, characters, and timelines to provide
//! narrative weaving suggestions for the writer.

use moling_core::error::{AppError, AppResult};
use moling_db::dao::chapter_dao::ChapterDao;
use moling_db::dao::project_dao::ProjectDao;
use moling_db::dao::vault_dao::VaultDao;
use sea_orm::{DatabaseConnection, Set};
use serde_json::Value as Json;

/// Business logic for narrative weaving operations.
#[derive(Clone)]
pub struct WeaveService {
    project_dao: ProjectDao,
    chapter_dao: ChapterDao,
    vault_dao: VaultDao,
}

impl WeaveService {
    pub fn new() -> Self {
        Self {
            project_dao: ProjectDao,
            chapter_dao: ChapterDao,
            vault_dao: VaultDao,
        }
    }

    /// List narrative weave patterns available in the system.
    pub async fn patterns(&self) -> AppResult<Json> {
        let patterns = vec![
            serde_json::json!({
                "id": "parallel",
                "name": "双线并行",
                "description": "两条故事线同时推进，在关键时刻交汇",
                "difficulty": "medium",
            }),
            serde_json::json!({
                "id": "flashback",
                "name": "倒叙穿插",
                "description": "通过倒叙揭示背景故事，增强叙事深度",
                "difficulty": "high",
            }),
            serde_json::json!({
                "id": "multi_pov",
                "name": "多视角叙事",
                "description": "通过多个角色的视角展开故事",
                "difficulty": "high",
            }),
            serde_json::json!({
                "id": "foreshadowing",
                "name": "伏笔网络",
                "description": "建立伏笔之间的关联网络，增强回收满足感",
                "difficulty": "medium",
            }),
            serde_json::json!({
                "id": "spiral",
                "name": "螺旋上升",
                "description": "每次回到相似主题但层级更高，情感递进",
                "difficulty": "medium",
            }),
            serde_json::json!({
                "id": "mosaic",
                "name": "马赛克拼图",
                "description": "多个看似无关的片段最终拼成完整故事",
                "difficulty": "high",
            }),
        ];
        Ok(serde_json::json!({
            "patterns": patterns,
            "count": patterns.len(),
        }))
    }

    /// Get weave-based narrative suggestions for a project.
    ///
    /// Analyzes chapters, characters, plots, and timelines to generate
    /// concrete weaving suggestions. Mirrors Python `WeaveService.get_suggestions`.
    pub async fn suggestions(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<Json> {
        let project = self.project_dao.find_by_id(db, project_id)
            .await?
            .ok_or_else(AppError::project_not_found)?;

        let chapters = self.chapter_dao.find_by_project(db, project_id).await?;
        let characters = self.vault_dao.find_characters(db, project_id).await?;
        let plot_promises = self.vault_dao.find_plot_promises(db, project_id).await?;
        let timelines = self.vault_dao.find_timeline_events(db, project_id).await?;

        let mut suggestions = Vec::new();

        // Suggestion 1: Plot thread integration
        if !plot_promises.is_empty() && chapters.len() > 3 {
            let dormant: Vec<_> = plot_promises.iter().filter(|p| p.status == "dormant").collect();
            if !dormant.is_empty() {
                suggestions.push(serde_json::json!({
                    "id": "weave_plot_threads",
                    "type": "plot_thread",
                    "priority": "high",
                    "title": "情节线索整合",
                    "description": format!("{} 条休眠伏笔可以编织进当前叙事", dormant.len()),
                    "suggestion": "在后续章节中逐步回收休眠伏笔，增强故事连贯性",
                    "affected_chapters": chapters.iter().rev().take(3).map(|c| c.id.clone()).collect::<Vec<_>>(),
                }));
            }
        }

        // Suggestion 2: Character arc weaving
        if characters.len() > 3 && chapters.len() > 5 {
            let low_appearance: Vec<_> = characters
                .iter()
                .filter(|c| c.chapter_count < chapters.len() as i32 / 3)
                .collect();
            if !low_appearance.is_empty() {
                suggestions.push(serde_json::json!({
                    "id": "weave_character_arcs",
                    "type": "character_arc",
                    "priority": "medium",
                    "title": "角色弧光编织",
                    "description": format!("{} 个角色出场率偏低，可加强其叙事线", low_appearance.len()),
                    "suggestion": "为低频角色设计支线剧情，与主线交织",
                    "affected_characters": low_appearance.iter().map(|c| c.name.clone()).collect::<Vec<_>>(),
                }));
            }
        }

        // Suggestion 3: Timeline coherence
        if timelines.len() > 1 {
            let chapter_nums: Vec<i32> = timelines.iter().map(|t| t.chapter_number).collect();
            let mut sorted = chapter_nums.clone();
            sorted.sort();
            if chapter_nums != sorted {
                suggestions.push(serde_json::json!({
                    "id": "weave_timeline",
                    "type": "timeline",
                    "priority": "high",
                    "title": "时间线调整",
                    "description": "检测到时间线事件顺序不一致",
                    "suggestion": "重新排序时间线事件，确保叙事逻辑清晰",
                }));
            }
        }

        // Suggestion 4: Pacing balance
        if chapters.len() > 5 {
            let word_counts: Vec<i32> = chapters.iter().map(|c| c.word_count).collect();
            if !word_counts.is_empty() {
                let avg: f64 = word_counts.iter().sum::<i32>() as f64 / word_counts.len() as f64;
                let variance: f64 = word_counts.iter()
                    .map(|w| {
                        let diff = *w as f64 - avg;
                        diff * diff
                    })
                    .sum::<f64>() / word_counts.len() as f64;
                let std_dev = variance.sqrt();

                if std_dev > avg * 0.5 {
                    suggestions.push(serde_json::json!({
                        "id": "weave_pacing",
                        "type": "pacing",
                        "priority": "medium",
                        "title": "节奏均衡调整",
                        "description": "章节长度波动较大，影响阅读节奏",
                        "suggestion": "保持每章长度相对均匀，将高/低潮章节交替安排",
                        "avg_words": avg.round(),
                        "std_dev": std_dev.round(),
                    }));
                }
            }
        }

        // Suggestion 5: Structure recommendation
        if !chapters.is_empty() && chapters.len() < 10 {
            suggestions.push(serde_json::json!({
                "id": "weave_structure",
                "type": "structure",
                "priority": "low",
                "title": "结构扩展建议",
                "description": "当前章节数较少，可规划更多叙事层次",
                "suggestion": "考虑加入支线剧情或次要角色视角，丰富叙事结构",
            }));
        }

        let overview = format!(
            "项目《{}》当前有 {} 章、{} 个角色、{} 个情节元素。",
            project.title,
            chapters.len(),
            characters.len(),
            plot_promises.len(),
        );

        Ok(serde_json::json!({
            "project_id": project_id,
            "suggestions": suggestions,
            "suggestion_count": suggestions.len(),
            "overview": overview,
            "analyzed_at": chrono::Utc::now().to_rfc3339(),
        }))
    }

    /// Apply a weave pattern to a project.
    ///
    /// Records the applied pattern on target chapters for traceability.
    pub async fn apply(
        &self,
        db: &DatabaseConnection,
        pattern: Json,
        target_chapter_ids: &[String],
    ) -> AppResult<Json> {
        let pattern_id = pattern
            .get("id")
            .and_then(|v| v.as_str())
            .unwrap_or("unknown");
        let pattern_name = pattern
            .get("name")
            .and_then(|v| v.as_str())
            .unwrap_or("未命名模式");

        let mut applied_to = Vec::new();
        for chapter_id in target_chapter_ids {
            if let Some(ch) = self.chapter_dao.find_by_id(db, chapter_id).await? {
                use sea_orm::{ActiveModelTrait, IntoActiveModel};
                let existing_prompt = ch.generation_prompt.clone();
                let mut active = ch.into_active_model();
                let note = format!("[Weave Applied: {pattern_id}] {pattern_name}");
                let new_prompt = if let Some(ref existing) = existing_prompt {
                    format!("{existing}\n{note}")
                } else {
                    note
                };
                active.generation_prompt = Set(Some(new_prompt));
                active.update(db).await.map_err(|e| {
                    AppError::internal(format!("Apply weave to chapter failed: {e}"))
                })?;
                applied_to.push(chapter_id.clone());
            }
        }

        tracing::info!(
            pattern_id,
            chapters = applied_to.len(),
            "Weave pattern applied"
        );

        Ok(serde_json::json!({
            "message": format!("已应用编织模式「{}」到 {} 个章节", pattern_name, applied_to.len()),
            "pattern_id": pattern_id,
            "applied_count": applied_to.len(),
            "applied_to": applied_to,
        }))
    }

    /// Deep analysis of narrative structure for coherence.
    ///
    /// Analyzes plot threads, character arcs, timeline consistency,
    /// and unresolved promises. Mirrors Python `WeaveService.analyze_project`.
    pub async fn analyze(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<Json> {
        let project = self.project_dao.find_by_id(db, project_id)
            .await?
            .ok_or_else(AppError::project_not_found)?;

        let chapters = self.chapter_dao.find_by_project(db, project_id).await?;
        let characters = self.vault_dao.find_characters(db, project_id).await?;
        let plot_promises = self.vault_dao.find_plot_promises(db, project_id).await?;
        let timelines = self.vault_dao.find_timeline_events(db, project_id).await?;

        // Plot threads analysis
        let plot_threads: Vec<Json> = plot_promises
            .iter()
            .map(|p| {
                let completion = match p.status.as_str() {
                    "resolved" => 1.0,
                    "active" => 0.5,
                    "dormant" => 0.1,
                    "abandoned" => 0.0,
                    _ => 0.3,
                };
                serde_json::json!({
                    "id": p.id,
                    "name": &p.description,
                    "status": p.status,
                    "completion": completion,
                })
            })
            .collect();

        // Character arcs analysis
        let total_chapters = chapters.len() as f64;
        let character_arcs: Vec<Json> = characters
            .iter()
            .map(|c| {
                let progress = if total_chapters > 0.0 {
                    (c.chapter_count as f64 / total_chapters).min(1.0)
                } else {
                    0.0
                };
                let arc_type = match c.role.as_str() {
                    "protagonist" => "growth",
                    "antagonist" => "rise_and_fall",
                    _ => "supporting",
                };
                serde_json::json!({
                    "character_name": c.name,
                    "role": c.role,
                    "arc_type": arc_type,
                    "progress": (progress * 10.0).round() / 10.0,
                    "chapter_count": c.chapter_count,
                })
            })
            .collect();

        // Timeline consistency
        let timeline_issues: Vec<String> = {
            let mut issues = Vec::new();
            let numbers: Vec<i32> = timelines.iter().map(|t| t.chapter_number).collect();
            let mut sorted = numbers.clone();
            sorted.sort();
            if numbers != sorted {
                issues.push("时间线事件存在章节顺序不一致".into());
            }
            if timelines.is_empty() {
                issues.push("未建立时间线事件".into());
            }
            issues
        };

        let timeline_score = if timelines.is_empty() {
            5.0
        } else if timeline_issues.is_empty() {
            8.0
        } else {
            6.0
        };

        // Unresolved promises
        let unresolved_promises: Vec<Json> = plot_promises
            .iter()
            .filter(|p| p.status != "resolved" && p.status != "abandoned")
            .map(|p| {
                serde_json::json!({
                    "id": p.id,
                    "description": p.description,
                    "status": p.status,
                    "priority": if p.status == "dormant" { "high" } else { "medium" },
                })
            })
            .collect();

        Ok(serde_json::json!({
            "project_id": project_id,
            "project_title": project.title,
            "analysis": {
                "plot_threads": plot_threads,
                "character_arcs": character_arcs,
                "timeline_consistency": {
                    "score": timeline_score,
                    "issues": timeline_issues,
                    "suggestions": if !timeline_issues.is_empty() {
                        vec!["修复时间线事件顺序".to_owned()]
                    } else {
                        vec!["时间线一致，可继续丰富事件细节".to_owned()]
                    },
                },
                "unresolved_promises": unresolved_promises,
                "summary": {
                    "total_chapters": chapters.len(),
                    "total_characters": characters.len(),
                    "total_plot_promises": plot_promises.len(),
                    "total_timeline_events": timelines.len(),
                    "unresolved_promise_count": unresolved_promises.len(),
                },
            },
            "created_at": chrono::Utc::now().to_rfc3339(),
        }))
    }

    /// Quick weave health score for a project (0.0 - 10.0).
    pub async fn health_score(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
    ) -> AppResult<f64> {
        let chapters = self.chapter_dao.count_by_project(db, project_id).await?;
        let characters = self.vault_dao.count_characters(db, project_id).await?;
        let promises = self.vault_dao.count_plot_promises(db, project_id).await?;

        if chapters == 0 {
            return Ok(0.0);
        }

        let mut score: f64 = 5.0; // baseline

        // Bonus: chapter count
        if chapters >= 5 { score += 1.0; }
        if chapters >= 20 { score += 0.5; }

        // Bonus: character variety
        if characters >= 3 { score += 0.5; }
        if characters >= 10 { score += 0.5; }

        // Bonus: plot complexity
        if promises >= 3 { score += 0.5; }
        if promises >= 10 { score += 0.5; }

        // Penalty: too few plot promises relative to chapters
        if chapters > 10 && promises < 3 {
            score -= 1.0;
        }

        // Cap at 10.0
        Ok(score.clamp(0.0, 10.0))
    }
}

impl Default for WeaveService {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_weave_service_constructs() {
        let _ = WeaveService::new();
    }

    #[test]
    fn test_patterns_returns_data() {
        // patterns() doesn't need DB, we can call it synchronously
        // We just verify the struct constructs
    }
}
